consumer/consumer.go

```diff
@@ -170,43 +187,59 @@ func (c *Consumer) handleMessages(ch <-chan *sarama.ConsumerMessage, offset offs
 	c.Logger.Println("partition messages - reading" + logSuffix)
 
 	for msg := range ch {
-		var attempts int
-
-		for {
-			attempts++
-
-			// Note: The second returned value is not checked because we will never receive messages for a topic
-			// that it does not have a handler for.
-			h, _ := c.handlers.Get(msg.Topic)
-
-			// if an error occurs during an unformat call, it usually means
-			// the chosen converter was the wrong one. the consumption must block
-			// and an error must be reported.
-			m, err := h.Converter.FromKafka(msg)
-			if err == nil {
-				err = h.Handler.HandleMessage(m)
-				if err == nil {
-					offset.MarkOffset(msg, "")
-					if c.Metrics != nil {
-						c.Metrics.Report(*m, &Metrics{
-							Attempts:        attempts,
-							RemainingOffset: max.HighWaterMarkOffset() - msg.Offset,
-						})
-					}
-					break
-				}
-			}
-
-			c.Logger.Printf("an error occured while consuming a message: %v, retrying after %s.", err, c.config.RetryInterval)
+		m, attempts := c.handleMsg(msg)
+		if m == nil {
+			c.Logger.Println("partition messages - closing" + logSuffix)
+			break
+		}
+		// We succeeded in sending the message.
+		offset.MarkOffset(msg, "")
+		if c.Metrics != nil {
+			c.Metrics.Report(*m, &Metrics{
+				Attempts:        attempts,
+				RemainingOffset: max.HighWaterMarkOffset() - msg.Offset,
+			})
+		}
+	}
+}
 
-			select {
-			case <-time.After(c.config.RetryInterval):
-			case <-c.quit:
-				c.Logger.Println("partition messages - closing" + logSuffix)
-				return
-			}
+// handleMsg attempts to handle the message. It retries until the
+// message can be handled OK or the consumer is stopped.
```

Comment:
:tulip: :tulip:

```suggestion
// message handler returns a nil error or the consumer is stopped.
```

Reply:
Hmm, I'm not quite sure, because it's not *just* that the handler has to return a nil error, but also that the message can be converted OK. Thanks for the suggestion, but I think I'll leave it as is for now.

Reply:
Indeed, I overlooked that part